src/misc/unicode.cpp

```diff
@@ -451,712 +653,878 @@ static bool wide_to_code_page(const std::vector<uint16_t> &str_in,
-
-                // Invalid grapheme that is not added (overridden) is OK here;
-                // at least CP 1258 definition from Unicode.org contains mapping
-                // of code page characters to combining marks, which is fine for
-                // converting texts, but a no-no for DOS emulation (where the
-                // number of output characters has to match the number of input
-                // characters).
-                // For such code page definitions, just override problematic
-                // mappings in the main mapping configuration file.
-                if (add_if_not_mapped(new_mapping, character_code, grapheme) &&
-                    !check_grapheme_valid(grapheme, file_name, line_num))
-                    return false;
-            }
-        } else {
-            error_parsing(file_name, line_num);
-            return false;
-        }
-    }
-
-    if (!check_import_status(in_file, file_name, new_mapping.empty()))
-        return false;
-
-    // Reading/parsing succeeded - use all the data read from file
-    mapping = new_mapping;
-    return true;
+	// Import main configuration file, telling how to construct UTF-8
+	// mappings for each and every supported code page
+
+	// Open the file
+	auto in_file = open_mapping_file(path_root, file_name_main);
+	if (!in_file)
+		return;
+
+	// Read and parse
+	bool file_empty      = true;
+	std::string line_str = " ";
+	size_t line_num      = 0;
+
+	uint16_t curent_code_page = 0;
+
+	config_mappings_t new_config_mappings;
+	config_duplicates_t new_config_duplicates;
+	config_aliases_t new_config_aliases;
+
+	while (get_line(in_file, line_str, line_num)) {
+		std::vector<std::string> tokens;
+		if (!get_tokens(line_str, tokens))
+			continue; // empty line
+		uint8_t character_code = 0;
+
+		if (GCC_UNLIKELY(tokens[0] == "ALIAS")) {
+			if ((tokens.size() != 3 && tokens.size() != 4) ||
+			    (tokens.size() == 4 && tokens[3] != "BIDIRECTIONAL")) {
+				error_parsing(file_name_main, line_num);
+				return;
+			}
+
+			uint16_t code_point_1 = 0;
+			uint16_t code_point_2 = 0;
+			if (!get_hex_16bit(tokens[1], code_point_1) ||
+			    !get_hex_16bit(tokens[2], code_point_2)) {
+				error_parsing(file_name_main, line_num);
+				return;
+			}
+
+			new_config_aliases.emplace_back(
+			        std::make_pair(code_point_1, code_point_2));
+
+			if (tokens.size() == 4) // check if bidirectional
+				new_config_aliases.emplace_back(
+				        std::make_pair(code_point_2, code_point_1));
+
+			curent_code_page = 0;
+
+		} else if (GCC_UNLIKELY(tokens[0] == "CODEPAGE")) {
+			auto check_no_code_page = [&](const uint16_t code_page) {
+				if ((new_config_mappings.find(code_page) !=
+				             new_config_mappings.end() &&
+				     new_config_mappings[code_page].valid) ||
+				    new_config_duplicates.find(code_page) !=
+				            new_config_duplicates.end()) {
+					error_code_page_defined(file_name_main,
+					                        line_num);
```

Comment:
Formatting (left alignment) in `if` criteria seem to be off

Reply:
- 8 character long tabulation
- 80 characters line limit
- policy of using very long, descriptive variable names
- C++17, where you often have to write very long statements to use modern library features
and the clang-format script produced something like this. Reformatted the code manually.

Reply:
(oh, and I see these were formatted inside the nested parentheses; might have already been OK in that regard! Sorry if this comment wasn't helpful!)